Reader: Add fixed table-layout to tables within full post view

alexjgustafson opened the pull request:
Fixes #986 , to keep tables from overflowing within Reader's full post view in firefox

Testing instructions:

1. Publish a new post including a table with images in two 50% width colums
2. View the post in the Reader
Result: the table is not constrained to the content area when the post is viewed in the Reader using Firefox.

Before:
<img width="1436" alt="before" src="https://cloud.githubusercontent.com/assets/933450/11565939/4d3907a2-99ae-11e5-9e29-913d3947ccb5.png">
<img width="336" alt="before-mobile" src="https://cloud.githubusercontent.com/assets/933450/11565940/4f0c438c-99ae-11e5-90db-0d5c7161bde0.png">

After:
<img width="1425" alt="after" src="https://cloud.githubusercontent.com/assets/933450/11565945/52ae698e-99ae-11e5-8b3c-bfef33655669.png">
<img width="332" alt="after-mobile" src="https://cloud.githubusercontent.com/assets/933450/11565946/550655e8-99ae-11e5-9a86-0b8bd8939ea6.png">

cc: @blowery @designsimply

blowery commented on client/reader/full-post/_style.scss:
```diff
@@ -217,6 +217,10 @@
 	}
 
+	table {
+		table-layout: fixed;
```
That seems... dangerous: https://developer.mozilla.org/en-US/docs/Web/CSS/table-layout

> Under the "fixed" layout method, the entire table can be rendered once the first table row has been downloaded and analyzed. This can speed up rendering time over the "automatic" layout method, but subsequent cell content may not fit in the column widths provided. Any cell that has content that overflows uses the overflow property to determine whether to clip the overflow content.

alexjgustafson commented:
The setting defaults to `auto`, with that in place for our test case, the photos are the widest unbreakable content in their respective columns. The table element has a `width:100%;` style in place, and each td has a width of 50% in place, but because the photos are the widest unbreakable content the photos set their respective columns width regardless of the table's width. Because of that, it's stretching out the table's cells past the containing `.reader__full-post-content` div, even though the table has those widths set.

It should be noted this effect occurs in other environments, not just the Reader. Here's the sample HTML provided in the issue in a post in the Twenty Fourteen theme:

<img width="1439" alt="blog___923_is_numbered__not_lettered___soon_we_ll_be_spelling_everything_with_letters" src="https://cloud.githubusercontent.com/assets/933450/11573505/b538a418-99d4-11e5-9da0-c7fcbfe90bec.png">

The Twenty Fifteen theme handles this issue by using `table-layout:fixed` for all table elements. Here's the sample HTML provided in the issue in a post in the Twenty Fifteen theme:

<img width="1437" alt="overflow_display_in_firefox___923_is_numbered__not_lettered" src="https://cloud.githubusercontent.com/assets/933450/11573435/5bf533d0-99d4-11e5-86d6-27d6626cc819.png">

Setting table-layout to `fixed` fixes the issue because the table renders based off the widths set in the tables first row rather than the contents within the table's cells. The content's width then gets handled within the cell as a container, following any overflow rules.

To produce the same effect without using `table-layout: fixed` I suppose we'd need to make adjustments to the content itself when found within a table. Images seem the most likely element to cause this width issue.  [Non-wrapping text would cause this effect too, but that happens even when the table-layout is fixed, and seems we handle it in the Reader view already](https://cloudup.com/cFJzYgwb3SV). A table nested within a table may also cause this effect with the right content.

I'll attempt a solution that at least accounts for the image case, and test out how other elements react to see if fixing more than images is necessary.

blowery commented:
The odd thing is that is _only_ does it in firefox. I wonder if this is a vague bit of the spec, or if Firefox is getting it wrong..

alexjgustafson commented:
I was hoping I could figure out what it is that Chrome and Firefox handle differently by using the same table setup without all the other styles already in Calypso, including normalize/reset styles that are in play. So I started working with this HTML on its own: https://gist.github.com/alexjgustafson/cf3fe38df0d15e52a675

When viewing that HTML alone, the images overflow the table in both Chrome and Firefox, regardless of `table-layout`. So I think you're right, there must be some other styles we already have in place that Chrome is supporting and Firefox just isn't. Having a hard time tracking down exactly what those are though.

blowery commented:
@alexjgustafson Have you had a chance to look at this again?

alexjgustafson commented:
@blowery a couple times I've attempted to find other solutions but I'm afraid I'm stumped beyond my original PR.

lancewillett commented:
Closing this one as no movement in several months — feel free to re-open if it's actionable and can be pushed in.
